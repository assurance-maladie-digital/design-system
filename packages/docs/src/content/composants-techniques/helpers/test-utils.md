---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
