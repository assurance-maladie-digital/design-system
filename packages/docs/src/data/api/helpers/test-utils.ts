import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
