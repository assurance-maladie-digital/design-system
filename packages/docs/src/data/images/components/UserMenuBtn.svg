<svg width="541" height="357" viewBox="0 0 541 357" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect width="541" height="357" fill="#E7ECF5"/>
<rect x="63" y="88" width="415" height="181" fill="white"/>
<circle cx="364" cy="179" r="20.8333" fill="#EDEDED"/>
<path d="M364 179C366.302 179 368.167 177.135 368.167 174.833C368.167 172.531 366.302 170.667 364 170.667C361.698 170.667 359.833 172.531 359.833 174.833C359.833 177.135 361.698 179 364 179ZM364 181.083C361.219 181.083 355.667 182.479 355.667 185.25V187.333H372.333V185.25C372.333 182.479 366.781 181.083 364 181.083Z" fill="#0084B2"/>
<rect opacity="0.5" x="158" y="181" width="170" height="19" rx="9.5" fill="#CED9EB"/>
<rect opacity="0.5" x="243" y="158" width="85" height="19" rx="9.5" fill="#CED9EB"/>
</svg>
