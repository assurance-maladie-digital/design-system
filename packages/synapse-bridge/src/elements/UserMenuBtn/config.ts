import { tokens } from "@cnamts/design-tokens";

export const config = {
	menu: {
		offsetY: true,
		location: "bottom left",
		zIndex: 4,
		minWidth: "198px",
		maxWidth: "198px",
		nudgeRight: 0,
	},
	btn: {
		variant: "text",
		height: "auto",
	},
	icon: {
		color: tokens.colors.cyan.darken20,
	},
	logoutListItem: {
		class: "text-primary",
		minWidth: "198px",
	},
	logoutIcon: {
		color: "primary",
		class: "mr-4",
	},
};
