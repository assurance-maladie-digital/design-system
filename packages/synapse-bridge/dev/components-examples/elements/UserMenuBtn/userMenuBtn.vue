<template>
	<PageContainer>
		<UserMenuBtn
			additional-information="Information supplémentaire"
			full-name="Virginie Beauchesne"
		/>
		<br/><br/>
		<UserMenuBtn full-name="Édith Cartier">
			<VListItem
				v-for="(item, index) in items"
				:key="index"
			>
				<VListItemTitle>{{ item }}</VListItemTitle>
			</VListItem>
		</UserMenuBtn>
		<br/><br/>
		<UserMenuBtn
			additional-information="Information supplémentaire"
			full-name="Virginie Beauchesne"
			mobile-version
		/>
		<br/><br/>
		<UserMenuBtn
			:vuetify-options="vuetifyOptions"
			full-name="Édith Cartier"
		>
			<VListItem
				v-for="(item, index) in items"
				:key="index"
			>
				<VListItemTitle>{{ item }}</VListItemTitle>
			</VListItem>
		</UserMenuBtn>
	</PageContainer>
</template>

<script lang="ts">
import UserMenuBtn from "@/elements/UserMenuBtn";
import PageContainer from "@/elements/PageContainer";
import { defineComponent } from "vue";

export default defineComponent({
	components: {
		UserMenuBtn,
		PageContainer,
	},
	data() {
		return {
			items: [
				'Administration',
				'Profil',
				'Paramètres'
			],
			vuetifyOptions: {
				menu: {
					minWidth: '264px',
					maxWidth: '264px'
				},
				btn: {
					variant: 'outlined',
				},
				icon: {
					color: 'primary'
				},
				logoutIcon: {
					class: 'd-none'
				}
			}
		};
	},
});
</script>
