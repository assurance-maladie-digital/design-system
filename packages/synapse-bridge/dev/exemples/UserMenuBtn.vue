<template>
	<PageContainer>
		<UserMenuBtn
			additional-information="Information supplémentaire"
			full-name="Virginie Beauchesne"
		/>
		<br/><br/>
		<UserMenuBtn full-name="Édith Cartier">
			<VListItem
				v-for="(item, index) in items"
				:key="index"
			>
				<VListItemTitle>{{ item }}</VListItemTitle>
			</VListItem>
		</UserMenuBtn>
	</PageContainer>
</template>

<script lang="ts">
import UserMenuBtn from "../src/elements/UserMenuBtn";
import PageContainer from "../src/elements/PageContainer";
import { defineComponent } from "vue";

export default defineComponent({
	components: {
		UserMenuBtn,
		PageContainer,
	},
	data() {
		return {
			items: [
				'Administration',
				'Profil',
				'Paramètres'
			]
		};
	},
});
</script>
