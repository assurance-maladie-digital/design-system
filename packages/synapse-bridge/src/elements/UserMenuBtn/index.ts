import UserMenuBtn from './UserMenuBtn.vue';

export default UserMenuBtn;
