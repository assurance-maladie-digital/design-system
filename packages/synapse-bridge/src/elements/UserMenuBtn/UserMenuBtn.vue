<script lang="ts">
import { defineComponent } from "vue";

import { mdiAccount, mdiLoginVariant } from '@mdi/js';

import { customizable } from '@/mixins/customizable';
import type { Refs } from '@/types';

import { config } from './config';
import { locales } from './locales';

export default defineComponent({
	mixins: [customizable(config)],
	props: {
		fullName: {
			type: String,
			required: true
		},
		additionalInformation: {
			type: String,
			default: undefined
		},
		label: {
			type: String,
			default: locales.label
		},
		hideUserIcon: {
			type: Boolean,
			default: false
		},
		hideLogoutBtn: {
			type: Boolean,
			default: false
		},
		mobileVersion: {
			type: Boolean,
			default: false
		}
	},
	data() {
		return {
			$refs: {} as Refs<{
				btn?: HTMLButtonElement;
			}>,
			locales,
			userIcon: mdiAccount,
			logoutIcon: mdiLoginVariant,
		};
	},
	computed: {
		btnPadding(): string {
      		return this.hideUserIcon ? 'pa-1 pa-sm-2' : this.isMobileVersion ? 'pa-0' : 'pa-1 pa-sm-3';
		},
		hasListContent(): boolean {
			return Boolean(this.$slots.default || !this.hideLogoutBtn);
		},
		isMobileVersion(): boolean {
      		return this.mobileVersion || ['xs', 'sm'].includes(this.$vuetify.display.name);
		},
		isMobileWithIcon(): boolean {
			return this.isMobileVersion && !this.hideUserIcon;
		}
	}
});
</script>

<template>
	<div class="vd-user-menu-btn-ctn d-inline-block">
		<VMenu
			v-bind="options.menu"
			:disabled="!hasListContent"
			transition="fade-transition"
			location="bottom end"
			z-index="9999"
		>
			<template #activator="{ props }">
				<VBtn
					v-bind="{
						...props,
						...options.btn
					}"
					:class="btnPadding"
					:icon="isMobileWithIcon"
					:size="isMobileWithIcon ? 'x-large' : 'default'"
					:height="isMobileWithIcon ? undefined : 'auto'"
					class="vd-user-menu-btn"
				>
					<span class="d-sr-only">
						{{ label }}
					</span>

					<span
						v-if="!isMobileVersion"
						class="d-flex flex-column align-end"
					>
						<span class="font-weight-bold">
							{{ fullName }}
						</span>

						<span class="text-grey text-darken-2 font-weight-medium">
							{{ additionalInformation }}
						</span>
					</span>

					<span
						v-if="isMobileVersion && hideUserIcon"
						class="font-weight-bold text-sm-caption"
					>
						{{ fullName }}
					</span>

					<slot name="icon">
						<VIcon
							v-if="!hideUserIcon"
							v-bind="options.icon"
							:class="isMobileVersion ? 'ml-0 pa-2' : 'ml-2 pa-2'"
							:size="isMobileWithIcon ? 'x-large' : 'default'"
							class="vd-user-icon"
						>
							{{ userIcon }}
						</VIcon>
					</slot>
				</VBtn>
			</template>

			<slot name="content">
				<VList
					v-if="hasListContent"
					v-bind="options.list"
				>
					<slot />

					<slot name="logout-item">
						<VListItem
							v-if="!hideLogoutBtn"
							v-bind="options.logoutListItem"
							class="logout"
							@click="$emit('logout')"
						>

							<div class="d-flex">
								<VIcon v-bind="options.logoutIcon" class="mr-4">
									{{ logoutIcon }}
								</VIcon>

								<VListItemTitle>
									{{ locales.logout }}
								</VListItemTitle>
							</div>
						</VListItem>
					</slot>
				</VList>
			</slot>
		</VMenu>
	</div>
</template>

<style lang="scss" scoped>
	@import '@cnamts/design-tokens/dist/tokens';
	.vd-user-menu-btn-ctn {
		position: relative;
	}
	.vd-user-menu-btn {
		outline: none;
		&:hover:before {
			background: #000;
			opacity: 0.05;
		}
		&:focus:before {
			background: $vd-primary;
			opacity: 0.08;
		}
		&:focus {
			background: rgba($vd-primary, 0.08) !important;
		}
	}
	:deep(.vd-user-menu-btn:focus > .v-btn__overlay) {
		opacity: 0 !important;
	}
	.vd-user-icon {
		width: 40px;
		height: 40px;
		background: $vd-grey-lighten-90;
		border-radius: 50%;
		svg, .v-icon__svg {
			width: 24px;
			height: 24px;
		}
	}
</style>
