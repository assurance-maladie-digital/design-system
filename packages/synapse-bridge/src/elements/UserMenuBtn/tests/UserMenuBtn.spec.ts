import { describe, it, expect } from 'vitest';
import { shallowMount, mount } from '@vue/test-utils';
import { vuetify } from '@tests/unit/setup';

import UserMenuBtn from '../';

describe('UserMenuBtn', () => {
	it("renders correctly with props", () => {
		const wrapper = shallowMount(UserMenuBtn, {
			props: {
				fullName: "Firstname Lastname",
				label: "Label",
				hideUserIcon: true,
				hideLogoutBtn: true,
				mobileVersion: true,
				additionalInformation: "Additional information",
			},
			global: {
				plugins: [vuetify],
			},
			slots: {
				default: "<p>the menu</p>",
			},
		});

		expect(wrapper).toMatchSnapshot();
	});

	it("renders btnPadding correctly", () => {
		const wrapper = mount(UserMenuBtn, {
			props: {
				fullName: "Firstname Lastname",
				mobileVersion: true,
				additionalInformation: "Additional information",
			},
			global: {
				plugins: [vuetify],
			},
		});

		expect(wrapper.vm.btnPadding).toBe("pa-0");
	});

	it("renders btnPadding correctly", () => {
		const wrapper = mount(UserMenuBtn, {
			props: {
				fullName: "Firstname Lastname",
				hideUserIcon: true,
				additionalInformation: "Additional information",
			},
			global: {
				plugins: [vuetify],
			},
		});

		expect(wrapper.vm.btnPadding).toBe("pa-1 pa-sm-2");
	});

	it("renders hasListContent correctly", () => {
		const wrapper = mount(UserMenuBtn, {
			props: {
				fullName: "Firstname Lastname",
				hideLogoutBtn: true,
			},
			global: {
				plugins: [vuetify],
			},
		});

		expect(wrapper.vm.hasListContent).toBe(false);
	});

	it("renders isMobileVersion computed correctly", () => {
		const wrapper = mount(UserMenuBtn, {
			props: {
				fullName: "Firstname Lastname",
				mobileVersion: true,
			},
			global: {
				plugins: [vuetify],
			},
		});

		expect(wrapper.vm.isMobileVersion).toBe(true);
	});

	it("renders isMobileWithIcon computed correctly", () => {
		const wrapper = mount(UserMenuBtn, {
			props: {
				fullName: "Firstname Lastname",
				mobileVersion: true,
				hideUserIcon: false,
			},
			global: {
				plugins: [vuetify],
			},
		});

		expect(wrapper.vm.isMobileWithIcon).toBe(true);
	});

	it("renders the full name correctly", () => {
		const wrapper = mount(UserMenuBtn, {
			props: {
				fullName: "Firstname Lastname",
				hideUserIcon: true,
				mobileVersion: false,
			},
			global: {
				plugins: [vuetify],
			},
		});

		expect(wrapper.find(".text-sm-caption").exists()).toBe(false);
	});

	it("renders the full name correctly in mobile mode", () => {
		const wrapper = mount(UserMenuBtn, {
			props: {
				fullName: "Firstname Lastname",
				hideUserIcon: true,
				mobileVersion: true,
			},
			global: {
				plugins: [vuetify],
			},
		});

		expect(wrapper.find(".text-sm-caption").exists()).toBe(true);
	});

	it("Display the menu on click", async () => {
		const wrapper = mount(UserMenuBtn, {
			props: {
				fullName: "Firstname Lastname",
			},
			global: {
				plugins: [vuetify],
			},
			slots: {
				default: "<p class='menu'>the menu</p>",
			},
		});

		await wrapper.find(".v-btn").trigger("click");
		expect(wrapper.find(".v-btn").attributes()["aria-expanded"]).toBe(
			"true"
		);
		expect(wrapper.find(".menu").exists());
	});

	it("Hide the menu on click", async () => {
		const wrapper = mount(UserMenuBtn, {
			props: {
				fullName: "Firstname Lastname",
			},
			global: {
				plugins: [vuetify],
			},
			slots: {
				default: "<p class='menu'>the menu</p>",
			},
		});

		await wrapper.find(".v-btn").trigger("click");
		await wrapper.find(".v-btn").trigger("click");
		expect(wrapper.find(".v-btn").attributes()["aria-expanded"]).toBe(
			"false"
		);
		expect(wrapper.find(".menu").exists()).toBe(false);
	});
});
