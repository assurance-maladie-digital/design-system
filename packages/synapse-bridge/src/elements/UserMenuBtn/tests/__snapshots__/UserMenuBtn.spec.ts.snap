// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserMenuBtn > renders correctly with props 1`] = `
<div data-v-3b944ebf="" class="vd-user-menu-btn-ctn d-inline-block">
  <v-menu-stub data-v-3b944ebf="" minwidth="198px" maxwidth="198px" attach="false" closeonback="true" contained="false" disabled="false" noclickanimation="false" modelvalue="false" persistent="false" scrim="false" zindex="9999" activatorprops="[object Object]" openonhover="false" closeoncontentclick="true" closedelay="250" opendelay="300" eager="false" locationstrategy="connected" location="bottom end" origin="auto" scrollstrategy="reposition" transition="fade-transition" offsety="true" nudgeright="0"></v-menu-stub>
</div>
`;
